src/scitokens_internal.cpp

```diff
@@ -284,29 +341,57 @@ std::string
 rs256_from_coords(const std::string &e_str, const std::string &n_str) {
     auto e_decode = b64url_decode_nopadding(e_str);
     auto n_decode = b64url_decode_nopadding(n_str);
+    std::unique_ptr<BIO, decltype(&BIO_free_all)> pubkey_bio(BIO_new(BIO_s_mem()), BIO_free_all);
     std::unique_ptr<BIGNUM, decltype(&BN_free)> e_bignum(BN_bin2bn(reinterpret_cast<const unsigned char *>(e_decode.c_str()), e_decode.size(), nullptr), BN_free);
     std::unique_ptr<BIGNUM, decltype(&BN_free)> n_bignum(BN_bin2bn(reinterpret_cast<const unsigned char *>(n_decode.c_str()), n_decode.size(), nullptr), BN_free);
-
-    std::unique_ptr<RSA, decltype(&RSA_free)> rsa(RSA_new(), RSA_free);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
-    rsa->e = e_bignum.get();
-    rsa->n = n_bignum.get();
-    rsa->d = nullptr;
+	
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    OSSL_PARAM *params;
+    std::unique_ptr<EVP_PKEY_CTX, decltype(&EVP_PKEY_CTX_free)> rsa_ctx(EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL), EVP_PKEY_CTX_free);
+	if (!rsa_ctx.get()) {
+	    throw UnsupportedKeyException("Failed to set RSA PKEY context");
+	}
+	
+    std::unique_ptr<OSSL_PARAM_BLD, decltype(&OSSL_PARAM_BLD_free)> param_build(OSSL_PARAM_BLD_new(), OSSL_PARAM_BLD_free);
+	if (!param_build.get()
+        || !OSSL_PARAM_BLD_push_BN_pad(param_build.get(),"e",e_bignum.get(),BN_num_bytes(e_bignum.get()))
+        || !OSSL_PARAM_BLD_push_BN_pad(param_build.get(),"n",n_bignum.get(),BN_num_bytes(n_bignum.get()))
+        || (params = OSSL_PARAM_BLD_to_param(param_build.get())) == NULL) {
+            throw UnsupportedKeyException("Failed to build RSA public key parameters");
+    }
+	
+    EVP_PKEY *pkey = NULL;
+	if (EVP_PKEY_fromdata_init(rsa_ctx.get()) <= 0
+	    || EVP_PKEY_fromdata(rsa_ctx.get(),&pkey,EVP_PKEY_PUBLIC_KEY,params) <= 0
+	    || pkey == NULL) {
+	        throw UnsupportedKeyException("Failed to set the RSA public key");
+	}
+	
+    if (PEM_write_bio_PUBKEY(pubkey_bio.get(), pkey) == 0) {
+        throw UnsupportedKeyException("Failed to serialize RSA public key");
+    }
+    EVP_PKEY_free(pkey);
+	OSSL_PARAM_free(params);
```

Comment:
Odd indention

Reply:
That one was an auto tab from return due to my text editor. Will fix.